src/log-store/src/kafka/producer.rs

```diff
@@ -238,24 +80,56 @@ impl OrderedBatchProducer {
         }
     }
 
+    fn set_running(&self, val: bool) {
+        self.running.store(val, Ordering::Relaxed);
+    }
+
     /// Writes `data` to the [`OrderedBatchProducer`].
     ///
     /// Returns the [ProduceResultHandle], which will receive a result when data has been committed to Kafka
     /// or an unrecoverable error has been encountered.
     ///
     /// ## Panic
     /// Panic if any [Record]'s `approximate_size` > `max_batch_bytes`.
-    pub(crate) async fn produce(&self, batch: Vec<Record>) -> Result<ProduceResultHandle> {
-        let receiver = {
-            let (tx, rx) = oneshot::channel();
-            self.sender
-                .send(ProduceRequest { batch, sender: tx })
-                .await
-                .context(error::SendProduceRequestSnafu)?;
-            rx
-        };
+    pub(crate) async fn produce(
+        &self,
+        region_id: RegionId,
+        batch: Vec<Record>,
+    ) -> Result<ProduceResultHandle> {
+        let (req, handle) = WorkerRequest::new_produce_request(region_id, batch);
+        if self.sender.send(req).await.is_err() {
+            warn!("OrderedBatchProducer is already exited");
+            self.set_running(false);
```

Comment:
There's no need to set `running` to false here. When `sender.send` returns an error, the corresponding recver must have already been dropped.